proxy/src/main/java/net/md_5/bungee/BungeeBossBar.java

```diff
@@ -0,0 +1,276 @@
+package net.md_5.bungee;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.EnumSet;
+import java.util.List;
+import java.util.UUID;
+
+import com.google.common.base.Preconditions;
+import com.google.common.collect.ImmutableList;
+import lombok.EqualsAndHashCode;
+import lombok.Getter;
+import lombok.ToString;
+import net.md_5.bungee.api.boss.BossBarColor;
+import net.md_5.bungee.api.boss.BossBarDivision;
+import net.md_5.bungee.api.boss.BossBarFlag;
+import net.md_5.bungee.api.chat.BaseComponent;
+import net.md_5.bungee.api.connection.ProxiedPlayer;
+import net.md_5.bungee.chat.ComponentSerializer;
+import net.md_5.bungee.protocol.DefinedPacket;
+import net.md_5.bungee.protocol.ProtocolConstants;
+import net.md_5.bungee.protocol.packet.BossBar;
+
+@ToString
+@EqualsAndHashCode
+public class BungeeBossBar implements net.md_5.bungee.api.boss.BossBar
+{
+
+    @Getter
+    private BaseComponent[] title;
+    @Getter
+    private BossBarColor color;
+    @Getter
+    private BossBarDivision division;
+    @Getter
+    private float health;
+    @Getter
+    private boolean visible;
+
+    private List<ProxiedPlayer> players;
+    private EnumSet<BossBarFlag> flags;
+
+    @ToString.Exclude
+    private UUID uuid;
+
+    public BungeeBossBar(BaseComponent[] title, BossBarColor color, BossBarDivision division, float health)
+    {
+        this.title = title;
+        this.color = color;
+        this.division = division;
+        this.health = health;
+        this.visible = true;
+        this.players = new ArrayList<>();
+        this.flags = EnumSet.noneOf( BossBarFlag.class );
+        this.uuid = UUID.randomUUID();
+    }
+
+    @Override
+    public void addPlayer(ProxiedPlayer player)
+    {
+        Preconditions.checkNotNull( player, "player" );
+        if ( !players.contains( player ) )
+        {
+            players.add( player );
+        }
+        if ( player.isConnected() && visible )
+        {
+            player.unsafe().sendPacket( addPacket() );
+        }
+    }
+
+    @Override
+    public void addPlayers(Iterable<ProxiedPlayer> players)
+    {
+        Preconditions.checkNotNull( players, "players" );
+        for ( ProxiedPlayer player : players )
+        {
+            addPlayer( player );
+        }
+    }
+
+    @Override
+    public void removePlayer(ProxiedPlayer player)
+    {
+        Preconditions.checkNotNull( player, "player" );
+        players.remove( player );
```

Comment:
Check if in list first

Reply:
The check is reudant as if the player does not contain, the method will do nothing except returning false

Reply:
Maybe move the remove into the if statement; a packet will still be sent nonetheless

Reply:
The return value will stop the execution method, right now ill leave it like that

Reply:
What return value?

Reply:
See latest changes

Reply:
You are using the ArrayList players as a Set here, so I suggest you actually make it a Set. Then you can replace the contains-add and contaisn-remove calls with just add and remove (in a Set, add returns true, if the element was not in the Set before and remove returns true if the element was in the Set before)

Reply:
> You are using the ArrayList players as a Set here, so I suggest you actually make it a Set. Then you can replace the contains-add and contaisn-remove calls with just add and remove (in a Set, add returns true, if the element was not in the Set before and remove returns true if the element was in the Set before)

Right now i check it. Can please see the updated code ok ty

Reply:
> > You are using the ArrayList players as a Set here, so I suggest you actually make it a Set. Then you can replace the contains-add and contaisn-remove calls with just add and remove (in a Set, add returns true, if the element was not in the Set before and remove returns true if the element was in the Set before)
> 
> Right now i check it. Can please see the updated code ok ty

You did not change anything about that.

Reply:
https://github.com/SpigotMC/BungeeCord/pull/2683/files#diff-55e1ca98f9181d0ccce7704aa488fbcbR89
If this isn't a check then kill me

Reply:
> https://github.com/SpigotMC/BungeeCord/pull/2683/files#diff-55e1ca98f9181d0ccce7704aa488fbcbR89
> If this isn't a check then kill me
I know that you check it before, but using a List as a Set is more error prone than using a Set directly.